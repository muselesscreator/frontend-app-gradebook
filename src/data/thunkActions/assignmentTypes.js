/* eslint-disable import/prefer-default-export */
import { StrictDict } from 'utils';
import actions from 'data/actions';
import selectors from 'data/selectors';
import LmsApiService from 'data/services/LmsApiService';

const { fetching, gotGradesFrozen } = actions.assignmentTypes;
const { gotBulkManagementConfig } = actions.config;

export const fetchAssignmentTypes = () => (
  (dispatch, getState) => {
    dispatch(fetching.started());
    return LmsApiService.fetchAssignmentTypes(selectors.app.courseId(getState()))
      .then(response => response.data)
      .then((data) => {
        dispatch(fetching.received(Object.keys(data.assignment_types)));
        dispatch(gotGradesFrozen(data.grades_frozen));
        dispatch(gotBulkManagementConfig(data.can_see_bulk_management));
      })
      .catch(() => {
        dispatch(fetching.error());
      });
  }
);

export default StrictDict({ fetchAssignmentTypes });
