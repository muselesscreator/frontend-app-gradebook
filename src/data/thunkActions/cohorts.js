/* eslint-disable import/prefer-default-export */

import cohorts from '../actions/cohorts';

import LmsApiService from '../services/LmsApiService';

const fetchCohorts = courseId => (
  (dispatch) => {
    dispatch(cohorts.fetching.started());
    return LmsApiService.fetchCohorts(courseId)
      .then(response => response.data)
      .then((data) => {
        dispatch(cohorts.fetching.received(data.cohorts));
      })
      .catch(() => {
        dispatch(cohorts.fetching.error());
      });
  }
);

export { fetchCohorts };
