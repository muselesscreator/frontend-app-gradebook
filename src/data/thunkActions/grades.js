import { getAuthenticatedHttpClient } from '@edx/frontend-platform/auth';

import grades from '../actions/grades';
import { sortAlphaAsc } from '../actions/utils';

import selectors from '../selectors';

import GRADE_OVERRIDE_HISTORY_ERROR_DEFAULT_MSG from '../constants/errors';

import LmsApiService from '../services/LmsApiService';

const {
  formatMaxAssignmentGrade,
  formatMinAssignmentGrade,
  formatMaxCourseGrade,
  formatMinCourseGrade,
  formatGradeOverrideForDisplay,
} = selectors.grades;

const defaultAssignmentFilter = 'All';

const fetchBulkUpgradeHistory = courseId => (
  // todo add loading effect
  dispatch => LmsApiService.fetchGradeBulkOperationHistory(courseId).then(
    (response) => { dispatch(grades.bulkHistory.received(response)); },
  ).catch(() => dispatch(grades.bulkHistory.error()))
);

const fetchGrades = (
  courseId,
  cohort,
  track,
  assignmentType,
  options = {},
) => (
  (dispatch, getState) => {
    dispatch(grades.fetching.started());
    const {
      assignment,
      assignmentGradeMax: assignmentMax,
      assignmentGradeMin: assignmentMin,
      courseGradeMin,
      courseGradeMax,
      includeCourseRoleMembers,
    } = selectors.filters.allFilters(getState());
    const { id: assignmentId } = assignment || {};
    const assignmentGradeMax = formatMaxAssignmentGrade(assignmentMax, { assignmentId });
    const assignmentGradeMin = formatMinAssignmentGrade(assignmentMin, { assignmentId });
    const courseGradeMinFormatted = formatMinCourseGrade(courseGradeMin);
    const courseGradeMaxFormatted = formatMaxCourseGrade(courseGradeMax);
    return LmsApiService.fetchGradebookData(
      courseId,
      options.searchText || null,
      cohort,
      track,
      {
        assignment: assignmentId,
        assignmentGradeMax,
        assignmentGradeMin,
        courseGradeMin: courseGradeMinFormatted,
        courseGradeMax: courseGradeMaxFormatted,
        includeCourseRoleMembers,
      },
    )
      .then(response => response.data)
      .then((data) => {
        dispatch(grades.received({
          grades: data.results.sort(sortAlphaAsc),
          cohort,
          track,
          assignmentType,
          prev: data.previous,
          next: data.next,
          courseId,
          totalUsersCount: data.total_users_count,
          filteredUsersCount: data.filtered_users_count,
        }));
        dispatch(grades.fetching.finished());
        if (options.showSuccess) {
          dispatch(grades.banner.open());
        }
      })
      .catch(() => {
        dispatch(grades.fetching.error());
      });
  }
);

const fetchGradeOverrideHistory = (subsectionId, userId) => (
  dispatch => LmsApiService.fetchGradeOverrideHistory(subsectionId, userId)
    .then(response => response.data)
    .then((data) => {
      if (data.success) {
        dispatch(grades.overrideHistory.received({
          overrideHistory: formatGradeOverrideForDisplay(data.history),
          currentEarnedAllOverride: data.override ? data.override.earned_all_override : null,
          currentPossibleAllOverride: data.override ? data.override.possible_all_override : null,
          currentEarnedGradedOverride: data.override ? data.override.earned_graded_override : null,
          currentPossibleGradedOverride: data.override
            ? data.override.possible_graded_override : null,
          originalGradeEarnedAll: data.original_grade ? data.original_grade.earned_all : null,
          originalGradePossibleAll: data.original_grade ? data.original_grade.possible_all : null,
          originalGradeEarnedGraded: data.original_grade
            ? data.original_grade.earned_graded : null,
          originalGradePossibleGraded: data.original_grade
            ? data.original_grade.possible_graded : null,
        }));
      } else {
        dispatch(grades.overrideHistory.errorFetching(data.error_message));
      }
    })
    .catch(() => {
      dispatch(grades.overrideHistory.errorFetching(GRADE_OVERRIDE_HISTORY_ERROR_DEFAULT_MSG));
    })
);

const fetchMatchingUserGrades = (
  courseId,
  searchText,
  cohort,
  track,
  assignmentType,
  showSuccess,
  options = {},
) => {
  const newOptions = { ...options, searchText, showSuccess };
  return fetchGrades(courseId, cohort, track, assignmentType, newOptions);
};

const fetchPrevNextGrades = (endpoint, courseId, cohort, track, assignmentType) => (
  (dispatch) => {
    dispatch(grades.fetching.started());
    return getAuthenticatedHttpClient().get(endpoint)
      .then(response => response.data)
      .then((data) => {
        dispatch(grades.fetching.results({
          grades: data.results.sort(sortAlphaAsc),
          cohort,
          track,
          assignmentType,
          prev: data.previous,
          next: data.next,
          courseId,
          totalUsersCount: data.total_users_count,
          filteredUsersCount: data.filtered_users_count,
        }));
        dispatch(grades.fetching.finished());
      })
      .catch(() => {
        dispatch(grades.fetching.error());
      });
  }
);

const submitFileUploadFormData = (courseId, formData) => (
  (dispatch) => {
    dispatch(grades.csvUpload.started());
    return LmsApiService.uploadGradeCsv(courseId, formData).then(() => {
      dispatch(grades.csvUpload.finished());
      dispatch(grades.uploadOverride.success(courseId));
    }).catch((err) => {
      dispatch(grades.uploadOverride.failure(courseId, err));
      if (err.status === 200 && err.data.error_messages.length) {
        const { error_messages: errorMessages, saved, total } = err.data;
        return dispatch(grades.csvUpload.error({ errorMessages, saved, total }));
      }
      return dispatch(grades.csvUpload.error({ errorMessages: ['Unknown error.'] }));
    });
  }
);

const updateGrades = (courseId, updateData, searchText, cohort, track) => (
  (dispatch) => {
    dispatch(grades.update.request());
    return LmsApiService.updateGradebookData(courseId, updateData)
      .then(response => response.data)
      .then((data) => {
        dispatch(grades.update.success({ courseId, data }));
        dispatch(fetchMatchingUserGrades(
          courseId,
          searchText,
          cohort,
          track,
          defaultAssignmentFilter,
          true,
          { searchText },
        ));
      })
      .catch((error) => {
        dispatch(grades.update.failure({ courseId, error }));
      });
  }
);

const updateGradesIfAssignmentGradeFiltersSet = (
  courseId,
  cohort,
  track,
  assignmentType,
) => (dispatch, getState) => {
  const { filters: { assignmentGradeMin, assignmentGradeMax } } = getState();
  const hasAssignmentGradeFiltersSet = assignmentGradeMax || assignmentGradeMin;
  if (hasAssignmentGradeFiltersSet) {
    dispatch(fetchGrades(
      courseId,
      cohort,
      track,
      assignmentType,
    ));
  }
};

export {
  fetchBulkUpgradeHistory,
  fetchGrades,
  fetchGradeOverrideHistory,
  fetchMatchingUserGrades,
  fetchPrevNextGrades,
  submitFileUploadFormData,
  updateGrades,
  updateGradesIfAssignmentGradeFiltersSet,
};
