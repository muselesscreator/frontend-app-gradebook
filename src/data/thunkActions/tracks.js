/* eslint-disable import/prefer-default-export */
import tracks from '../actions/tracks';

import selectors from '../selectors';

import {
  fetchBulkUpgradeHistory,
} from './grades';

import LmsApiService from '../services/LmsApiService';

const fetchTracks = courseId => (
  (dispatch) => {
    dispatch(tracks.fetching.started());
    return LmsApiService.fetchTracks(courseId)
      .then(response => response.data)
      .then((data) => {
        dispatch(tracks.fetching.received(data.course_modes));
        if (selectors.tracks.hasMastersTrack(data.course_modes)) {
          dispatch(fetchBulkUpgradeHistory(courseId));
        }
      })
      .catch(() => {
        dispatch(tracks.fetching.error());
      });
  }
);

export {
  fetchTracks,
};
