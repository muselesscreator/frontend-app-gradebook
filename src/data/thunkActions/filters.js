/* eslint-disable import/prefer-default-export */

import * as filters from '../actions/filters';

import { getFilters } from '../selectors/filters';

import { fetchGrades } from './grades';

const updateIncludeCourseRoleMembers = (includeCourseRoleMembers) => (dispatch, getState) => {
  dispatch(filters.update.includeCourseRoleMembers(includeCourseRoleMembers));
  const state = getState();
  const { cohort, track, assignmentType } = getFilters(state);
  dispatch(fetchGrades(state.grades.courseId, cohort, track, assignmentType));
};

export {
  updateIncludeCourseRoleMembers,

};
