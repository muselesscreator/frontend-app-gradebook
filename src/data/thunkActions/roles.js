/* eslint-disable import/prefer-default-export */
import roles from '../actions/roles';
import selectors from '../selectors';

import { fetchCohorts } from './cohorts';
import {
  fetchGrades,
} from './grades';
import { fetchTracks } from './tracks';
import { fetchAssignmentTypes } from './assignmentTypes';

import LmsApiService from '../services/LmsApiService';

const allowedRoles = ['staff', 'instructor', 'support'];

const getRoles = courseId => (
  (dispatch, getState) => LmsApiService.fetchUserRoles(courseId)
    .then(response => response.data)
    .then((response) => {
      const isAllowedRole = (role) => (
        (role.course_id === courseId) && allowedRoles.includes(role.role)
      );

      const canUserViewGradebook = (response.is_staff || (response.roles.some(isAllowedRole)));
      dispatch(roles.received({ canUserViewGradebook, courseId }));
      const {
        cohort,
        track,
        assignmentType,
      } = selectors.filters.allFilters(getState());
      if (canUserViewGradebook) {
        dispatch(fetchGrades(courseId, cohort, track, assignmentType));
        dispatch(fetchTracks(courseId));
        dispatch(fetchCohorts(courseId));
        dispatch(fetchAssignmentTypes(courseId));
      }
    })
    .catch(() => {
      dispatch(roles.errorFetching());
    }));

export {
  getRoles,
};
