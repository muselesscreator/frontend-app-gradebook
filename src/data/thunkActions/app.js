import { StrictDict } from 'utils';

import actions from 'data/actions';
import { fetchGradeOverrideHistory } from './grades';
import { fetchRoles } from './roles';

export const setModalStateFromTable = ({ userEntry, subsection }) => (
  (dispatch) => {
    dispatch(fetchGradeOverrideHistory(subsection.module_id, userEntry.user_id));
    dispatch(actions.app.setModalStateFromTable({ subsection, userEntry }));
  }
);

export const initialize = (courseId, urlQuery) => (dispatch) => {
  dispatch(actions.app.setCourseId(courseId));
  dispatch(actions.filters.initialize(urlQuery));
  dispatch(fetchRoles());
};

export default StrictDict({
  initialize,
  setModalStateFromTable,
});
