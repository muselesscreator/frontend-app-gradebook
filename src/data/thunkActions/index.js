import { StrictDict } from 'utils';
import app from './app';
import assignmentTypes from './assignmentTypes';
import cohorts from './cohorts';
import filters from './filters';
import grades from './grades';
import roles from './roles';
import tracks from './tracks';

export default StrictDict({
  app,
  assignmentTypes,
  cohorts,
  filters,
  grades,
  roles,
  tracks,
});
