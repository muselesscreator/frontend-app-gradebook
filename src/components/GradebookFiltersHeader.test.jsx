import React from 'react';
import { shallow } from 'enzyme';

import thunkActions from 'data/thunkActions';

import {
  GradebookFiltersHeader,
  mapDispatchToProps,
} from './GradebookFiltersHeader';

jest.mock('@edx/paragon', () => ({
  Icon: jest.fn().mockName('Paragon.Icon'),
  IconButton: () => 'IconButton',
}));

jest.mock('@edx/paragon/icons', () => ({
  close: jest.fn().mockName('Paragon.icons.Close'),
}));

jest.mock('data/thunkActions', () => ({
  __esModule: true,
  default: {
    app: {
      filterMenu: {
        close: jest.fn().mockName('closeFilterMenu'),
      },
    },
  },
}));

describe('GradebookFiltersHeader', () => {
  let props;
  beforeEach(() => {
    props = {
      closeMenu: jest.fn().mockName('props.closeMenu'),
    };
  });

  describe('Component', () => {
    describe('snapshots', () => {
      test('basic snapshot', () => {
        const el = shallow(<GradebookFiltersHeader {...props} />);
        expect(el).toMatchSnapshot();
      });
    });
  });
  describe('mapDispatchToProps', () => {
    describe('closeMenu', () => {
      test('from thunkActions.app.filterMenu.close', () => {
        expect(mapDispatchToProps.closeMenu).toEqual(
          thunkActions.app.filterMenu.close,
        );
      });
    });
  });
});
