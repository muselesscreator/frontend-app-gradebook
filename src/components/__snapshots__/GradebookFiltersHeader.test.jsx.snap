// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GradebookFiltersHeader Component snapshots basic snapshot 1`] = `
<Fragment>
  <h2>
    <mockConstructor
      className="fa fa-filter"
    />
  </h2>
  <IconButton
    alt="Close Filters"
    aria-label="Close Filters"
    className="p-1"
    iconAs={[MockFunction Paragon.Icon]}
    onClick={[MockFunction props.closeMenu]}
  />
</Fragment>
`;
